// @ngx-pwa/local-storage@9.0.3/esm5/lib/storages/local-storage.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { mapTo, map } from 'rxjs/operators';
import { StorageMap } from './storage-map.service';
import * as i0 from "@angular/core";
import * as i1 from "./storage-map.service";
var LocalStorage = /** @class */ (function () {
    /* Use the `StorageMap` service to avoid code duplication */
    function LocalStorage(storageMap) {
        this.storageMap = storageMap;
    }
    Object.defineProperty(LocalStorage.prototype, "length", {
        /**
         * Number of items in storage wrapped in an `Observable`
         *
         * @example
         * this.localStorage.length.subscribe((length) => {
         *   console.log(length);
         * });
         */
        get: function () {
            return this.storageMap.size;
        },
        enumerable: true,
        configurable: true
    });
    LocalStorage.prototype.getItem = function (key, schema) {
        if (schema) {
            /* Backward compatibility with version <= 7 */
            var schemaFinal = ('schema' in schema) ? schema.schema : schema;
            return this.storageMap.get(key, schemaFinal).pipe(
            /* Transform `undefined` into `null` to align with `localStorage` API */
            map(function (value) { return (value !== undefined) ? value : null; }));
        }
        else {
            return this.storageMap.get(key).pipe(
            /* Transform `undefined` into `null` to align with `localStorage` API */
            map(function (value) { return (value !== undefined) ? value : null; }));
        }
    };
    /**
     * Set an item in storage.
     * Note that setting `null` or `undefined` will remove the item to avoid some browsers issues.
     * @param key The item's key
     * @param data The item's value
     * @param schema Optional JSON schema to validate the data
     * @returns A RxJS `Observable` to wait the end of the operation
     *
     * @example
     * this.localStorage.set('key', 'value').subscribe(() => {});
     */
    LocalStorage.prototype.setItem = function (key, data, schema) {
        return this.storageMap.set(key, data, schema).pipe(
        /* Transform `undefined` into `true` for backward compatibility with v7 */
        mapTo(true));
    };
    /**
     * Delete an item in storage
     * @param key The item's key
     * @returns A RxJS `Observable` to wait the end of the operation
     *
     * @example
     * this.localStorage.delete('key').subscribe(() => {});
     */
    LocalStorage.prototype.removeItem = function (key) {
        return this.storageMap.delete(key).pipe(
        /* Transform `undefined` into `true` for backward compatibility with v7 */
        mapTo(true));
    };
    /**
     * Delete all items in storage
     * @returns A RxJS `Observable` to wait the end of the operation
     *
     * @example
     * this.localStorage.clear().subscribe(() => {});
     */
    LocalStorage.prototype.clear = function () {
        return this.storageMap.clear().pipe(
        /* Transform `undefined` into `true` for backward compatibility with v7 */
        mapTo(true));
    };
    LocalStorage.ctorParameters = function () { return [
        { type: StorageMap }
    ]; };
    LocalStorage.ɵprov = i0.ɵɵdefineInjectable({ factory: function LocalStorage_Factory() { return new LocalStorage(i0.ɵɵinject(i1.StorageMap)); }, token: LocalStorage, providedIn: "root" });
    LocalStorage = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], LocalStorage);
    return LocalStorage;
}());
export { LocalStorage };
